File: csharp/HyperCast.Tests/CorpusTests.cs
using System.Text;
using System.Text.Json;

namespace HyperCast.Tests;

/// <summary>
/// Replays the shared conformance corpus (<c>corpus/*.json</c> at the repository root) —
/// the same files the Rust core's own test suite replays — through this binding's UTF-8
/// doors. Fault spans are byte offsets into the UTF-8 input, which is exactly what these
/// doors receive, so span assertions hold verbatim. This is the byte-for-byte polyglot
/// contract: a vector that fails here is a break in the promise, not just a failing test.
/// </summary>
public sealed class CorpusTests
{
	static readonly string _corpusDirectory = FindCorpusDirectory();

	static string FindCorpusDirectory()
	{
		for (var dir = new DirectoryInfo(AppContext.BaseDirectory); dir is not null; dir = dir.Parent)
		{
			var corpus = Path.Combine(dir.FullName, "corpus");
			if (Directory.Exists(corpus))
				return corpus;
		}
		throw new DirectoryNotFoundException($"corpus directory not found above {AppContext.BaseDirectory}");
	}

	static JsonElement[] Corpus(string name)
	{
		using var document = JsonDocument.Parse(File.ReadAllText(Path.Combine(_corpusDirectory, name)));
		return [.. document.RootElement.EnumerateArray().Select(vector => vector.Clone())];
	}

	static byte[] InputBytes(JsonElement vector) =>
		Encoding.UTF8.GetBytes(vector.GetProperty("input").GetString()!);

	static NumFormat FormatOf(JsonElement vector)
	{
		if (!vector.TryGetProperty("format", out var format))
			return NumFormat.Invariant;
		return new(
			format.GetProperty("decimal_sep").GetString()![0],
			format.GetProperty("group_sep").GetString()![0],
			(NumStyles)format.GetProperty("flags").GetUInt32());
	}

	static void AssertVerdict<T>(string domain, JsonElement vector, Verdict<T> verdict, Func<JsonElement, T> expected)
		where T : struct
	{
		var input = vector.GetProperty("input").GetString()!;
		var expect = vector.GetProperty("expect").GetString()!;
		// TryGetValue consumption, not case-type patterns: with an open generic T the union
		// matcher refuses to bind a pattern variable (CS8780) — and this is the hot-loop
		// consumption shape anyway. The compile-checked pattern form is exercised with
		// concrete case types in CastTests.
		if (verdict.TryGetValue(out Success<T> success))
		{
			expect.ShouldBe("ok", $"{domain}: '{input}' unexpectedly parsed");
			success.Value.ShouldBe(expected(vector), $"{domain}: '{input}'");
			return;
		}
		verdict.TryGetValue(out Fault fault).ShouldBeTrue($"{domain}: '{input}' produced no case");
		var reason = expect switch
		{
			"empty" => CastFailure.Empty,
			"malformed" => CastFailure.Malformed,
			"out_of_range" => CastFailure.OutOfRange,
			_ => throw new InvalidOperationException($"{domain}: '{input}' expected '{expect}' but faulted"),
		};
		fault.Reason.ShouldBe(reason, $"{domain}: '{input}'");
		if (vector.TryGetProperty("fault", out var span))
		{
			fault.Offset.ShouldBe(span[0].GetInt32(), $"{domain}: '{input}' fault offset");
			fault.Length.ShouldBe(span[1].GetInt32(), $"{domain}: '{input}' fault length");
		}
	}

	[Fact]
	void Boolean_corpus()
	{
		foreach (var vector in Corpus("boolean.json"))
			AssertVerdict("boolean", vector, Cast.Boolean(InputBytes(vector)),
				static v => v.GetProperty("value").GetBoolean());
	}

	[Fact]
	void Integer_corpus()
	{
		foreach (var vector in Corpus("integer.json"))
		{
			var input = InputBytes(vector);
			var format = FormatOf(vector);
			switch (vector.GetProperty("type").GetString())
			{
				case "i8":
					AssertVerdict("integer", vector, Cast.SByte(input, format), static v => v.GetProperty("value").GetSByte());
					break;
				case "i16":
					AssertVerdict("integer", vector, Cast.Int16(input, format), static v => v.GetProperty("value").GetInt16());
					break;
				case "i32":
					AssertVerdict("integer", vector, Cast.Int32(input, format), static v => v.GetProperty("value").GetInt32());
					break;
				case "i64":
					AssertVerdict("integer", vector, Cast.Int64(input, format), static v => v.GetProperty("value").GetInt64());
					break;
				case "u8":
					AssertVerdict("integer", vector, Cast.Byte(input, format), static v => v.GetProperty("value").GetByte());
					break;
				case "u16":
					AssertVerdict("integer", vector, Cast.UInt16(input, format), static v => v.GetProperty("value").GetUInt16());
					break;
				case "u32":
					AssertVerdict("integer", vector, Cast.UInt32(input, format), static v => v.GetProperty("value").GetUInt32());
					break;
				case "u64":
					AssertVerdict("integer", vector, Cast.UInt64(input, format), static v => v.GetProperty("value").GetUInt64());
					break;
				case var other:
					throw new InvalidOperationException($"integer: unknown type '{other}'");
			}
		}
	}

	[Fact]
	void Real_corpus()
	{
		foreach (var vector in Corpus("real.json"))
		{
			var input = InputBytes(vector);
			var format = FormatOf(vector);
			switch (vector.GetProperty("type").GetString())
			{
				case "f32":
					AssertVerdict("real", vector, Cast.Single(input, format),
						static v => (float)v.GetProperty("value").GetDouble());
					break;
				case "f64":
					AssertVerdict("real", vector, Cast.Double(input, format),
						static v => v.GetProperty("value").GetDouble());
					break;
				case var other:
					throw new InvalidOperationException($"real: unknown type '{other}'");
			}
		}
	}

	[Fact]
	void Uuid_corpus()
	{
		foreach (var vector in Corpus("uuid.json"))
			AssertVerdict("uuid", vector, Cast.Uuid(InputBytes(vector)), static v =>
			{
				var hex = v.GetProperty("value").GetString()!;
				var bytes = Convert.FromHexString(hex);
				return new Guid(bytes, bigEndian: true);
			});
	}

	static DateTimeOffset ExpectedInstant(JsonElement vector) =>
		DateTimeOffset.UnixEpoch
			.AddTicks(vector.GetProperty("seconds").GetInt64() * TimeSpan.TicksPerSecond
				+ vector.GetProperty("nanos").GetInt64() / 100);

	[Fact]
	void Timestamp_corpus()
	{
		foreach (var vector in Corpus("timestamp.json"))
			AssertVerdict("timestamp", vector, Cast.Timestamp(InputBytes(vector)), ExpectedInstant);
	}

	[Fact]
	void Unix_corpus()
	{
		foreach (var vector in Corpus("unix.json"))
		{
			var precision = (UnixPrecision)vector.GetProperty("precision").GetUInt32();
			AssertVerdict("unix", vector, Cast.Unix(InputBytes(vector), precision), ExpectedInstant);
		}
	}

	[Fact]
	void Date_corpus()
	{
		foreach (var vector in Corpus("date.json"))
			AssertVerdict("date", vector, Cast.Date(InputBytes(vector)), static v => new DateOnly(
				v.GetProperty("year").GetInt32(),
				v.GetProperty("month").GetInt32(),
				v.GetProperty("day").GetInt32()));
	}

	[Fact]
	void Time_corpus()
	{
		foreach (var vector in Corpus("time.json"))
			AssertVerdict("time", vector, Cast.Time(InputBytes(vector)),
				static v => new TimeOnly((long)(v.GetProperty("nanos").GetUInt64() / 100)));
	}

	[Fact]
	void Duration_corpus()
	{
		foreach (var vector in Corpus("duration.json"))
			AssertVerdict("duration", vector, Cast.Duration(InputBytes(vector)),
				static v => new TimeSpan(
					v.GetProperty("seconds").GetInt64() * TimeSpan.TicksPerSecond
					+ v.GetProperty("nanos").GetInt64() / 100));
	}
}

File: swift/Tests/HyperCastTests/CorpusTests.swift
import Foundation
import XCTest
@testable import HyperCast

/// Replays the shared conformance corpus (`corpus/*.json` at the repository root) — the
/// same files every other binding's suite replays — through this binding's `[UInt8]`
/// doors. Fault spans are byte offsets into the UTF-8 input, which is exactly what these
/// doors receive, so span assertions hold verbatim.
final class CorpusTests: XCTestCase {
    private static let corpusDirectory: URL = {
        var dir = URL(fileURLWithPath: #filePath)
        while dir.path != "/" {
            let candidate = dir.appendingPathComponent("corpus")
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate
            }
            dir.deleteLastPathComponent()
        }
        fatalError("corpus directory not found above \(#filePath)")
    }()

    private func corpus(_ name: String) throws -> [[String: Any]] {
        let data = try Data(contentsOf: Self.corpusDirectory.appendingPathComponent(name))
        return try JSONSerialization.jsonObject(with: data) as! [[String: Any]]
    }

    private func inputBytes(_ vector: [String: Any]) -> [UInt8] {
        Array((vector["input"] as! String).utf8)
    }

    private func format(of vector: [String: Any]) -> NumFormat {
        guard let format = vector["format"] as? [String: Any] else { return .invariant }
        return NumFormat(
            decimalSeparator: (format["decimal_sep"] as! String).unicodeScalars.first!,
            groupSeparator: (format["group_sep"] as! String).unicodeScalars.first!,
            styles: NumStyles(rawValue: UInt32(format["flags"] as! Int)))
    }

    private func assertVerdict<T: Equatable>(
        _ domain: String, _ vector: [String: Any], _ verdict: Verdict<T>, _ expected: T?,
        file: StaticString = #filePath, line: UInt = #line
    ) {
        let input = vector["input"] as! String
        let expect = vector["expect"] as! String
        // The union payoff, in the language where it's mandatory: two cases, no default.
        switch verdict {
        case .success(let value):
            XCTAssertEqual(expect, "ok", "\(domain): '\(input)' unexpectedly parsed", file: file, line: line)
            XCTAssertEqual(value, expected, "\(domain): '\(input)'", file: file, line: line)
        case .fault(let fault):
            let reason: CastFailure? = switch expect {
            case "empty": .empty
            case "malformed": .malformed
            case "out_of_range": .outOfRange
            default: nil
            }
            XCTAssertEqual(fault.reason, reason, "\(domain): '\(input)'", file: file, line: line)
            if let span = vector["fault"] as? [Int] {
                XCTAssertEqual(fault.offset, span[0], "\(domain): '\(input)' fault offset", file: file, line: line)
                XCTAssertEqual(fault.length, span[1], "\(domain): '\(input)' fault length", file: file, line: line)
            }
        }
    }

    func testBooleanCorpus() throws {
        for vector in try corpus("boolean.json") {
            assertVerdict("boolean", vector, try Cast.bool(inputBytes(vector)), vector["value"] as? Bool)
        }
    }

    func testIntegerCorpus() throws {
        for vector in try corpus("integer.json") {
            let input = inputBytes(vector)
            let format = format(of: vector)
            // JSONSerialization surfaces integers as NSNumber; u64::MAX survives via int64Value's
            // bit pattern, re-read unsigned where the door is unsigned.
            let number = vector["value"] as? NSNumber
            switch vector["type"] as! String {
            case "i8":
                assertVerdict("integer", vector, try Cast.i8(input, format: format), number?.int8Value)
            case "i16":
                assertVerdict("integer", vector, try Cast.i16(input, format: format), number?.int16Value)
            case "i32":
                assertVerdict("integer", vector, try Cast.i32(input, format: format), number?.int32Value)
            case "i64":
                assertVerdict("integer", vector, try Cast.i64(input, format: format), number?.int64Value)
            case "u8":
                assertVerdict("integer", vector, try Cast.u8(input, format: format), number?.uint8Value)
            case "u16":
                assertVerdict("integer", vector, try Cast.u16(input, format: format), number?.uint16Value)
            case "u32":
                assertVerdict("integer", vector, try Cast.u32(input, format: format), number?.uint32Value)
            case "u64":
                assertVerdict("integer", vector, try Cast.u64(input, format: format), number?.uint64Value)
            case let other:
                XCTFail("integer: unknown type \(other)")
            }
        }
    }

    func testRealCorpus() throws {
        for vector in try corpus("real.json") {
            let input = inputBytes(vector)
            let format = format(of: vector)
            let number = vector["value"] as? NSNumber
            switch vector["type"] as! String {
            case "f32":
                assertVerdict("real", vector, try Cast.f32(input, format: format),
                              number.map { Float($0.doubleValue) })
            case "f64":
                assertVerdict("real", vector, try Cast.f64(input, format: format), number?.doubleValue)
            case let other:
                XCTFail("real: unknown type \(other)")
            }
        }
    }

    func testUuidCorpus() throws {
        for vector in try corpus("uuid.json") {
            var expected: UUID?
            if let hex = vector["value"] as? String {
                var bytes = [UInt8]()
                var index = hex.startIndex
                while index < hex.endIndex {
                    let next = hex.index(index, offsetBy: 2)
                    bytes.append(UInt8(hex[index..<next], radix: 16)!)
                    index = next
                }
                expected = bytes.withUnsafeBytes { UUID(uuid: $0.load(as: uuid_t.self)) }
            }
            assertVerdict("uuid", vector, try Cast.uuid(inputBytes(vector)), expected)
        }
    }

    private func expectedInstant(_ vector: [String: Any]) -> Date? {
        guard let seconds = vector["seconds"] as? Int64 else { return nil }
        let nanos = vector["nanos"] as! Int64
        return Date(timeIntervalSince1970: Double(seconds) + Double(nanos) / 1_000_000_000)
    }

    func testTimestampCorpus() throws {
        for vector in try corpus("timestamp.json") {
            assertVerdict("timestamp", vector, try Cast.timestamp(inputBytes(vector)), expectedInstant(vector))
        }
    }

    func testUnixCorpus() throws {
        for vector in try corpus("unix.json") {
            let precision = UnixPrecision(rawValue: UInt32(vector["precision"] as! Int))!
            assertVerdict("unix", vector, try Cast.unix(inputBytes(vector), precision: precision),
                          expectedInstant(vector))
        }
    }

    func testExcelSerialCorpus() throws {
        for vector in try corpus("excel_serial.json") {
            let epoch = ExcelEpoch(rawValue: UInt32(vector["epoch"] as! Int))!
            assertVerdict("excel_serial", vector, try Cast.excelSerial(inputBytes(vector), epoch: epoch),
                          expectedInstant(vector))
        }
    }

    func testDateCorpus() throws {
        for vector in try corpus("date.json") {
            var expected: DateComponents?
            if let year = vector["year"] as? Int {
                expected = DateComponents(
                    year: year, month: (vector["month"] as! Int), day: (vector["day"] as! Int))
            }
            assertVerdict("date", vector, try Cast.date(inputBytes(vector)), expected)
        }
    }

    func testDateOrderCorpus() throws {
        for vector in try corpus("date_order.json") {
            guard let order = DateOrder(rawValue: UInt32(vector["order"] as! Int)) else {
                XCTFail("date_order: unknown order")
                continue
            }
            var expected: DateComponents?
            if let year = vector["year"] as? Int {
                expected = DateComponents(
                    year: year, month: (vector["month"] as! Int), day: (vector["day"] as! Int))
            }
            assertVerdict("date_order", vector, try Cast.date(inputBytes(vector), order: order), expected)
        }
    }

    func testDateTimeCorpus() throws {
        for vector in try corpus("datetime.json") {
            guard let order = DateOrder(rawValue: UInt32(vector["order"] as! Int)) else {
                XCTFail("datetime: unknown order")
                continue
            }
            var expected: DateComponents?
            if let year = vector["year"] as? Int {
                let nanos = UInt64(vector["nanos_of_day"] as! Int)
                let secondOfDay = nanos / 1_000_000_000
                expected = DateComponents(
                    year: year, month: (vector["month"] as! Int), day: (vector["day"] as! Int),
                    hour: Int(secondOfDay / 3_600),
                    minute: Int(secondOfDay % 3_600 / 60),
                    second: Int(secondOfDay % 60),
                    nanosecond: Int(nanos % 1_000_000_000))
            }
            assertVerdict("datetime", vector, try Cast.dateTime(inputBytes(vector), order: order), expected)
        }
    }

    func testTimeCorpus() throws {
        for vector in try corpus("time.json") {
            var expected: DateComponents?
            if let nanosOfDay = (vector["nanos"] as? NSNumber)?.uint64Value {
                let (secondOfDay, nano) = nanosOfDay.quotientAndRemainder(dividingBy: 1_000_000_000)
                let (hour, rest) = secondOfDay.quotientAndRemainder(dividingBy: 3_600)
                let (minute, second) = rest.quotientAndRemainder(dividingBy: 60)
                expected = DateComponents(
                    hour: Int(hour), minute: Int(minute), second: Int(second), nanosecond: Int(nano))
            }
            assertVerdict("time", vector, try Cast.time(inputBytes(vector)), expected)
        }
    }

    func testDurationCorpus() throws {
        for vector in try corpus("duration.json") {
            var expected: Duration?
            if let seconds = vector["seconds"] as? Int64 {
                expected = Duration.seconds(seconds) + .nanoseconds(vector["nanos"] as! Int64)
            }
            assertVerdict("duration", vector, try Cast.duration(inputBytes(vector)), expected)
        }
    }
}
